sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RedirectPolicy.java

```diff
@@ -0,0 +1,103 @@
+    private final RedirectStrategy redirectStrategy;
+
+    private Set<String> attemptedRedirectUrls = new HashSet<>();
+    private String redirectedEndpointUrl;
+
+    /**
+     * Creates {@link RedirectPolicy} with default {@link MaxAttemptRedirectStrategy} as {@link RedirectStrategy} and
+     * use the provided {@code statusCode} to determine if this request should be redirected
+     * and MAX_REDIRECT_ATTEMPTS for the try count.
+     */
+    public RedirectPolicy() {
+        this(new MaxAttemptRedirectStrategy(MAX_REDIRECT_ATTEMPTS));
+    }
+
+    /**
+     * Creates {@link RedirectPolicy} with default {@link MaxAttemptRedirectStrategy} as {@link RedirectStrategy} and
+     * use the provided {@code statusCode} to determine if this request should be redirected.
+     *
+     * @param redirectStrategy The {@link RedirectStrategy} used for redirection.
+     * @throws NullPointerException When {@code statusCode} is null.
+     */
+    public RedirectPolicy(RedirectStrategy redirectStrategy) {
+        this.redirectStrategy = Objects.requireNonNull(redirectStrategy, "'redirectStrategy' cannot be null.");
+    }
+
+    @Override
+    public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+        return attemptRedirect(context, next, context.getHttpRequest(), 1);
+    }
+
+    /**
+     * Function to process through the HTTP Response received in the pipeline
+     * and redirect sending the request with new redirect url.
+     */
+    private Mono<HttpResponse> attemptRedirect(final HttpPipelineCallContext context,
+                                               final HttpPipelineNextPolicy next,
+                                               final HttpRequest originalHttpRequest,
+                                               final int redirectAttempt) {
+        // make sure the context is not modified during redirect, except for the URL
+        context.setHttpRequest(originalHttpRequest.copy());
+        if (this.redirectedEndpointUrl != null) {
+            context.getHttpRequest().setUrl(this.redirectedEndpointUrl);
+        }
+        return next.clone().process()
+            .flatMap(httpResponse -> {
+                String responseLocation =
+                    tryGetRedirectHeader(httpResponse.getHeaders(), redirectStrategy.getLocationHeader());
+                if (isValidRedirectStatusCode(httpResponse.getStatusCode()) &&
+                    isAllowedRedirectMethod(httpResponse.getRequest().getHttpMethod()) &&
+                    responseLocation != null &&
+                    redirectStrategy.shouldAttemptRedirect(responseLocation, redirectAttempt,
+                        redirectStrategy.getMaxAttempts(), attemptedRedirectUrls)) {
+                    attemptedRedirectUrls.add(responseLocation);
+                    this.redirectedEndpointUrl = responseLocation;
+                    return attemptRedirect(context, next, originalHttpRequest, redirectAttempt + 1);
```

Comment:
Before attempting the redirect the `HttpResponse` body will need to be drained, check out what the `RetryPolicy` currently does for this. Should also think about centralizing that logic into implementation somewhere.